### PR TITLE
gitserver: Pass pathspecs to git archive

### PR DESCRIPTION
**Pull request by chrismwendt**
Keegan [discovered](https://sourcegraph.slack.com/archives/CHXHX7XAS/p1647331744501559) that the "paths" passed to `git archive` are actually "pathspecs" with glob-like syntax. Git docs:

https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefpathspecapathspec

This PR:

- Renames `Paths` to `Pathspecs` to more accurately reflect that reality
- Adds a nominal type `type Pathspec string` to help avoid footgunning ourselves
- Prefixes pathspecs with `:(literal)` where appropriate to prevent `*` and `?` from being interpreted as special characters

CC

- @keegancsmith who discovered it
- @tsenart who worked on lockfile detection
- @mollylogue who has been working on sub-repo perms and git archive recently

## Test plan

Ran locally, did search-based code intel on a couple commits to exercise the `git archive` calls in the symbols service.

### CODE REVIEW

File: internal/codeintel/lockfiles/service_test.go [hunk 1]
```diff
@@ -19,27 +19,8 @@ import (
 func TestListDependencies(t *testing.T) {
 	ctx := context.Background()
 
-	t.Run("empty ls-files return", func(t *testing.T) {
```
Comment (keegancsmith) [inline internal/codeintel/lockfiles/service_test.go:22]: not clear to me why this was removed?

File: internal/gitserver/client.go [hunk 2]
```diff
@@ -258,11 +258,21 @@ func addrForKey(key string, addrs []string) string {
...
 
+// Pathspec is a git term for a pattern that matches paths using glob-like syntax.
+// https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefpathspecapathspec
+type Pathspec string
+
+// Literal matches a path without interpreting "*" or "?" as special characters.
+func Literal(s string) Pathspec { return Pathspec(":(literal)" + s) }
+
+// Suffix matches paths ending with the given suffix (useful for matching paths by basename).
+func Suffix(s string) Pathspec { return Pathspec("*" + s) }
```
Comment (keegancsmith) [inline internal/gitserver/client.go:274]: this API doesn't read that well at the callsites / look great in godoc. Maybe give these functions a prefix to somewhat namespace them / make it clear they are related.

**Review by keegancsmith — APPROVED** [review]
LGTM, thanks for fixing this. Yay tiny types :)

File: cmd/gitserver/server/server.go [hunk 3]
```diff
@@ -860,11 +860,11 @@ func (s *Server) handleRepoUpdate(w http.ResponseWriter, r *http.Request) {
...
-		q       = r.URL.Query()
-		treeish = q.Get("treeish")
-		repo    = q.Get("repo")
-		format  = q.Get("format")
-		paths   = q["path"]
+		q         = r.URL.Query()
+		treeish   = q.Get("treeish")
+		repo      = q.Get("repo")
+		format    = q.Get("format")
+		pathspecs = q["pathspec"]
```
[the remark this answers is not in the document]
Comment (chrismwendt) [reply cmd/gitserver/server/server.go]: Good catch, I reverted the query param back to `path`.

File: internal/codeintel/lockfiles/service_test.go (hunk 1 above)
Comment (chrismwendt) [reply internal/codeintel/lockfiles/service_test.go:22]: Same 🙃 Restored the test.

File: internal/gitserver/client.go (hunk 2 above)
Comment (chrismwendt) [reply internal/gitserver/client.go]: Added a prefix so now it's `PathspecLiteral()` and `PathspecSuffix()`, and tweaked the docstrings.

**Pull request merged**